submissions/krylenger/a-tiny-JS-world_post_OOP/index.js

```diff
@@ -0,0 +1,126 @@
+/* Refer to https://github.com/OleksiyRudenko/a-tiny-JS-world for the task details */
+
+const dog = {
+    species: "dog",
+  },
+  cat = {
+    species: "cat",
+  },
+  woman = {
+    species: "human",
+  },
+  man = {
+    species: "human",
+  },
+  catWoman = {
+    species: "catWoman",
+  };
+
+const capitalizeWord = (word) => {
+  return word.charAt(0).toUpperCase() + word.slice(1);
+}
+
+const createInhabitant = ({species}) => {
+  let self = {
+    species
+  };
+  return Object.assign(self);
+};
+
+const outputHumanIntroduction = ({
+  species,
+  name,
+  gender,
+  legs,
+  hands,
+  friends,
+}) => `Hi! My name is <em>${name}</em> and I'm <em>${species}</em>. 
+    My gender is <em>${gender}</em>. I have <em>${legs}</em> legs and <em>${hands}</em> hands. 
+    These are my friends: <em>${friends.join(", ")}</em>.<em> 
+    </em><br><br>`;
+
+const outputPetIntroduction = ({
+  word,
+  name,
+  gender,
+  legs,
+  paws,
+  friends,
+  species,
+}) => {
+  const capitalizedWord = capitalizeWord(word);
+
+  return `${capitalizedWord}! ${capitalizedWord} ${word} <em>${name}</em>. ${capitalizedWord} ${word} <em>${species}</em>. 
+  ${capitalizedWord} ${word} <em>${gender}</em>. ${capitalizedWord} <em>${legs}</em> ${word} <em>${paws}</em> ${word}. 
+  Meow ${word} ${word}: <em>${friends.join(", ")}</em>.<em> 
+  </em><br><br>`;
+};
+
+const humanSays = (self) => ({
+  introduction: () => {
+    return outputHumanIntroduction(self);
+  },
+});
+
+const carnivoraPetSays = (self) => ({
+  introduction: () => {    
+    return outputPetIntroduction(self);
+  },
+});
+
+const createCarnivoraPet = (speciesObj) => {
+  const self = createInhabitant(speciesObj);
+  self.legs = 4;
+  self.paws = 4;
+  return Object.assign(self, carnivoraPetSays(self));
+}
+
+const createHuman = (speciesObj) => {
+  const self = createInhabitant(speciesObj);
+  self.legs = 2;
+  self.hands = 2;
+  return Object.assign(self, humanSays(self));
+}
+
+const createCat = (name, gender, friends) => {
+  const self = createCarnivoraPet(cat);
+  self.word = 'meow';
+  return Object.assign(self, {name, gender, friends});
+};
+
+const createDog = (name, gender, friends) => {
+  const self = createCarnivoraPet(dog);
+  self.word = 'woof';
+  return Object.assign(self, {name, gender, friends});
+};
+
+const createMan = (name, friends) => {
+  const self = createHuman(man);
+  self.gender = "male";
+  return Object.assign(self, {name, friends});
+};
+
+const createWoman = (name, friends) => {
+  const self = createHuman(woman);
+  self.gender = "female";
+  return Object.assign(self, {name, friends});
+};
+
+const createCatWoman = (name, friends) => {
+  const self = createWoman(catWoman);
+  self.species = "catWoman";
```

Comment:
Why `const self = createWoman(catWoman);` is not sufficient?

Reply:
It's no sufficient because I instantiate from Woman class and don't have access to methods of Cat (e.g. output method) 
I suggest that solution is to compose catWoman from Inhabitant class and add all required methods and parameters. 
Is it a correct approach?

Reply:
Or even instantiate catWoman from Woman but also compose meower() and carnivoraPetSays to have all required properties and methods.